nexus/src/app/allowed_source_ips.rs

```diff
@@ -0,0 +1,147 @@
+        opctx: &OpContext,
+    ) -> Result<AllowedSourceIps, Error> {
+        self.db_datastore
+            .allowed_source_ips_view(opctx)
+            .await
+            .map(AllowedSourceIps::from)
+    }
+
+    /// Upsert the allowlist of source IPs that can reach user-facing services.
+    pub async fn allowed_source_ips_upsert(
+        &self,
+        opctx: &OpContext,
+        remote_addr: IpAddr,
+        params: params::AllowedSourceIpsUpdate,
+    ) -> Result<(), Error> {
+        if let external::AllowedSourceIps::List(list) = &params.allowed_ips {
+            // Size limits on the allowlist.
+            const MAX_ALLOWLIST_LENGTH: usize = 1000;
+            if list.len() > MAX_ALLOWLIST_LENGTH {
+                let message = format!(
+                    "Source IP allowlist is limited to {} entries, found {}",
+                    MAX_ALLOWLIST_LENGTH,
+                    list.len(),
+                );
+                return Err(Error::invalid_request(message));
+            }
+            if list.is_empty() {
+                return Err(Error::invalid_request(
+                    "Source IP allow list must not be empty",
+                ));
+            }
+
+            // Some basic sanity-checks on the addresses in the allowlist.
+            //
+            // The most important part here is checking that the source address
+            // the request came from is on the allowlist. This is our only real
+            // guardrail to prevent accidentally preventing any future access to
+            // the rack!
+            for entry in list.iter() {
+                if !entry.contains(remote_addr) {
+                    return Err(Error::invalid_request(
+                        "The source IP allow list would prevent access \
+                        from the current client! Ensure that the allowlist \
+                        contains an entry that continues to allow access \
+                        from this peer.",
+                    ));
+                }
+                if entry.ip().is_unspecified() {
+                    return Err(Error::invalid_request(
+                        "Source IP allowlist may not contain the \
+                        unspecified address. Use \"any\" to allow \
+                        any source to connect to user-facing services.",
+                    ));
+                }
+                if entry.prefix() == 0 {
+                    return Err(Error::invalid_request(
+                        "Source IP allowlist entries may not have \
+                        a netmask of /0.",
+                    ));
+                }
+            }
+        };
+
+        // Actually insert the new allowlist.
+        self.db_datastore
+            .allowed_source_ips_upsert(opctx, params.allowed_ips)
+            .await
+            .map(|_| ())?;
+
+        // Notify the sled-agents of the updated firewall rules.
+        match nexus_networking::plumb_service_firewall_rules(
```

Comment:
Failing here seems bad; the client will get a 500 that they don't understand, and won't know whether the rules they just supplied were applied. I guess we don't know either.

I don't have a better suggestion, though. Ideally this would trigger the firewall rule RPW, or we'd have one that would guarantee eventual correctness, but (a) we don't and (b) you probably still want to tell the client that these rules have actually been applied.

Could we construct a 500 that includes more information than the usual "no information" we get that suggests they retry their request or something? :-/

Reply:
I don't think we can add any more context to a 503, unfortunately. I can return a 400, but it's _not_ really a client error, we dropped the ball. I think a 503 is probably the best thing we can do at this point (along with an RPW). I'll make sure we return a 503 in this case.

Reply:
Hmm, that's unfortunate. I was thinking of `HttpError`, which we _can_ add more context to; e.g., in wicketd:

https://github.com/oxidecomputer/omicron/blob/faa8b8f7ed6eb0fcd6585dca862b028132a31282/wicketd/src/http_entrypoints.rs#L385-L392

It doesn't seem worth it to try to fix this on this PR - should I file an issue for adding more context to `Error`?

Reply:
I think that sounds worthwhile, yeah. I'm not sure why the error type doesn't expose control over that, when the `HttpError` we always convert it do does. Perhaps a way to enforce consistency? I don't know, but an issue sounds good.

Reply:
Opened https://github.com/oxidecomputer/omicron/issues/5692